File: CodingSamples/DotNet/Fundamentals/Generics/TypeParamTest/DemoApp2/Common/SimpleStack.cs
namespace DemoApp.Common;

//open generic type 
public class SimpleStack<E> : IStackReader<E>, IStackWriter<E>
{
    class Node
    {
        internal E element;
        internal Node below;
    }

    private Node top;

    public void Push(E item)
    {
        top = new Node { element = item, below = top };
    }

    public E Pop()
    {
        Node temp = top;
        top = top.below;
        return temp.element;
    }

    public bool Empty()
    {
        return top is null;
    }
}

File: CodingSamples/Java/Fundamentals/Generics/TypeParamTest/demo2/SimpleStack.java
class SimpleStack<E> {
    
    //inner member class
    class Node {

        E element;
        Node below;

        Node(E value) {
            element = value;
            below = top;
        }
    }

    private Node top;

    public void push(E item) {
        top = new Node(item);
    }

    public E pop() {
        Node temp = top;
        top = top.below;
        return temp.element;
    }

    public boolean empty() {
        return top == null;
    }
}
